### PR TITLE
Improved MCDC coverage

### PR DESCRIPTION
**Pull request by theyoucheng**
This patch aims to achieve complete mcdc coverage, which is not fulfilled previously. A set of test cases for it are also added.

### CODE REVIEW

File: src/goto-instrument/cover.cpp [hunk 1]
```diff
@@ -315,9 +350,456 @@ std::set<exprt> collect_mcdc_controlling(
...
+  {
+    exprt no_op(src);
+    no_op.make_not();
+    return not eval_exprt(atomic_exprts, no_op);
+  }
+  else if(is_condition(src))
+  {
+    return atomic_exprts.find(src)->second;
+  }
+}
```
Comment (kroening) [inline src/goto-instrument/cover.cpp]: There should be an else case here, returning something.

File: src/goto-instrument/cover.cpp [hunk 2]
```diff
@@ -315,9 +350,456 @@ std::set<exprt> collect_mcdc_controlling(
...
+{
+  // to obtain the set of atomic conditions
+  std::set<exprt> conditions;
+  for(auto &x: controlling)
+  {
+    std::set<exprt> new_conditions = collect_conditions(x);
+    conditions.insert(new_conditions.begin(), new_conditions.end());
+  }
+
+  while(true)
```
Comment (kroening) [inline src/goto-instrument/cover.cpp]: The control flow in this loop seems very obscure

File: src/goto-instrument/cover.cpp [hunk 3]
```diff
@@ -315,9 +350,456 @@ std::set<exprt> collect_mcdc_controlling(
...
+  Inputs: 
+
+ Outputs:
+
+ Purpose: To check if we can find the mcdc pair of the 
+          input ''exprt_set'' regarding the atomic exprt ''c''
+
+\*******************************************************************/
+
+bool find_mcdc_pair(
```
Comment (kroening) [inline src/goto-instrument/cover.cpp]: Rename to has_mcdc_pair

File: src/goto-instrument/cover.cpp [hunk 4]
```diff
@@ -315,9 +350,456 @@ std::set<exprt> collect_mcdc_controlling(
...
+  auto xit=atomic_exprts_e2.begin();
+  while(eit!=atomic_exprts_e1.end())
+  {
+    if(eit->second!=xit->second)
+    diff_count++;
+    if(diff_count>1) break;
+    eit++;
+    xit++;
+  }
+  if(diff_count==1) return true;
```
Comment (kroening) [inline src/goto-instrument/cover.cpp]: Formatting

File: src/goto-instrument/cover.cpp [hunk 5]
```diff
@@ -315,9 +350,456 @@ std::set<exprt> collect_mcdc_controlling(
...
+
+ Outputs: +1 : not negated
+           0 : not contained in E
+          -1 : negated
+
+ Purpose: The sign of exprt ''e'' within the super-exprt ''E''
+
+\*******************************************************************/
+
+void sign_of_exprt(const exprt &e, const exprt &E, std::set<signed> &signs)
```
Comment (kroening) [inline src/goto-instrument/cover.cpp]: Return the set as return value

File: src/goto-instrument/cover.cpp (hunk 1 above)
Comment (theyoucheng) [inline src/goto-instrument/cover.cpp]: Just committed changes. Please have a check. Thanks.

**Pull request merged**